### ISSUE TITLE
[Question] Below Expected Results Using BERTEmbedding

### ISSUE THREAD

**Issue by tnolan8** [BrikerMan/Kashgari#178]
A clear and concise description of what you want to know.
Hey!

I'm currently creating a classifier that categorises Tweets into one of 4 labels, ['anger', 'fear', 'joy', 'sadness']. I'm using the BertEmbedding combined with a BLSTMModel but not achieving very good results on my training data (accuracy is never over 0.55). I've only started using BERT recently but I'm wondering if there's something that I'm missing that's causing it to not perform as I'd expect. I've created similar classifiers using scikit-learn models such as SVM, NB ect and even an LSTM using Keras that utilises GloVe embeddings, all of which achieve >0.90 accuracy on my training data.

I'm starting to wonder if because the BERTEmbeddings are contextual that it might have something to do with the quantity of data that I have (~10K Tweets) not being enough. 

Any ideas what might be dragging down the accuracy/pointers in the right direction? Any help would be much appreciated, thank you!

(I can link a notebook also if that makes things easier)

**Comment by BrikerMan** [comment]
**Long story short**, you need to use the bert-tokenizer to tokenize your sentence, here is the [document](https://kashgari.bmio.net/embeddings/bert-embedding/#example-usage-text-classification)

**Detailed info**

Let's assume you have two samples like this. Let's see the difference between
- I did like apple
- I don't like apple

```python
# space tokenized result
space_tokenized = [
    ['He', 'did', 'like', 'apple'], 
    ['I', "doesn't", 'like', 'apple']
]

# bert tokenized result
bert_tokenized = [
    ['[CLS]', 'he', 'di', '##d', 'like', 'apple', '[SEP]'],
    ['[CLS]', 'i', 'do', '##es', '##n', "'", 't', 'like', 'apple', '[SEP]']
]

# Convert tokens to idx
space_tokenized_idx = [
  [ 101  100  100 8993 8350  102    0    0    0    0]
  [ 101  100  100 8993 8350  102    0    0    0    0]
]

bert_tokenized_idx = [
  [ 101  151 9796 8168 8993 8350  102    0    0    0]
  [ 101  151 8895 8333 8171  112  162 8993 8350  102]
]
```

You can see, when using space tokenize, it will convert to the same idx sequence. So it becomes an identical data for the model. This will lose lots of data.

**Comment by BrikerMan** [comment]
You also could try kashgari without BERT embedding~

**Comment by tnolan8** [comment]
Hey! Thanks for the very quick reply! This is more or less what I was trying:

```
embedding = BERTEmbedding('./moodryng/uncased_L-12_H-768_A-12/',
                   task=kashgari.CLASSIFICATION, 
                   sequence_length=128)`

train_df['input_x'] = [tokenizer.tokenize(sentence) for sentence in train_df['text'].values]
val_df['input_x'] = [tokenizer.tokenize(sentence) for sentence in val_df['text'].values]

train_x, train_y = list(train_df['input_x']), list(train_df['label'])
validate_x, validate_y = list(val_df['input_x']), list(val_df['label'])

model = BLSTMModel(embedding)
model.fit(train_x, train_y, validate_x, validate_y, epochs=8, batch_size=128)
```

This however never reaches >55% accuracy. I've also tried taking out the embedding and using just `nltk` as the tokenizer and then `model = BLSTMModel()` but the accuracy still isn't what I'd expect. I'll try do some more digging to see if I can find out what might be going wrong for me here!

**Comment by BrikerMan** [comment]
Just try the model without BERT first, you only need to change the one line.

```python
model = BLSTMModel()
```

**Comment by tnolan8** [comment]
I gave `model = BLSTMModel()` a try and I'm still not seeing great success. I created a Bidirectional LSTM in Keras using:

```
inp = Input(shape=(MAX_SEQUENCE_LENGTH,))
x = Embedding(MAX_NB_WORDS, EMBEDDING_DIM, input_length=X.shape[1], weights=[embedding_matrix])(inp)
x = Bidirectional(LSTM(100, return_sequences=True, dropout=0.25, recurrent_dropout=0.1))(x)
x = GlobalMaxPool1D()(x)
x = Dense(4, activation="softmax")(x)
model = Model(inputs=inp, outputs=x)
model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])
```

and then tried to re-create the same model using a custom model with Kashgari using:

```
def build_model_arc(self): 
    output_dim = len(self.pre_processor.label2idx)
    config = self.hyper_parameters
    embed_model = self.embedding.embed_model

    layers_seq = []
    layers_seq.append(L.Bidirectional(L.LSTM(100, return_sequences=True)))
    layers_seq.append(L.GlobalMaxPool1D())
    layers_seq.append(L.Dense(4, activation="softmax"))

    tensor = embed_model.output
    for layer in layers_seq:
        tensor = layer(tensor)

    self.tf_model = tf.keras.Model(embed_model.inputs, tensor)
```

Looking at the model summary's the only difference I can see is in number of parms going into the Embedding layer. ScreenshotA is the output of the top Model from Keras, ScreenshotB is the output from the custom model built with Kashgari.

ScreenshotA: http://prntscr.com/olpsow (Keras)
ScreenshotB: http://prntscr.com/olptgt (Kashgari)

I'm obviously missing something with regards to my input into the model or something like that but I can't tell what. I appreciate the help.

**Comment by BrikerMan** [comment]
I noticed the difference here, Kashgari uses `categorical_crossentropy` for multi-class classification and `binary_crossentropy` for multi-label classification. But you used the `binary_crossentropy` for multi-class classification. That might give you the wrong result. 
Still, you can customize the loss function by code below, then compare the differences.

```python
model =  BLSTMModel(embedding) 
# or
model =  BLSTMModel()
model.build_model(train_x, train_y, validate_x, validate_y, epochs=8, batch_size=128)
model.compile_model(loss='binary_crossentropy')
# Then fit
model.fit(train_x, train_y, validate_x, validate_y, epochs=8, batch_size=128)
```

**Comment by tnolan8** [comment]
Hey! So I figured it out! The accuracy has actually be fine the entire time, the accuracy I was seeing with my standard Keras model was completely wrong. You're right I should never have been using `binary_crossentropy`, it turns out that when you do that for a multi-class problem Keras isn't sure what to do when you specify `metrics=['accuracy']`. Very good run down of the issue here: https://stackoverflow.com/questions/42081257/keras-binary-crossentropy-vs-categorical-crossentropy-performance

Everything is all good now, thanks for your help :)

**Comment by Syauri** [comment]
How can I use bert-tokenizer for sequence labelling task? The documentation has an example for classification task.
